lib/src/arithmetic/ones_complement_adder.dart

```diff
@@ -12,76 +12,95 @@ import 'package:rohd/rohd.dart';
-/// ones-complement values.
+/// ones-complement values, producing a magnitude and sign.
 class OnesComplementAdder extends Adder {
   /// The sign of the result
   Logic get sign => output('sign');
 
-  /// The end-around carry which should be added to the resulting [sum]
-  /// If the input [carryOut] is not null, this value is stored there.
-  ///  Otherwise, the end-around carry is internally added to [sum]
-  Logic? get carryOut => tryOutput('carryOut');
+  /// The end-around carry which should be added to the resulting [sum].
+  /// If the input [endAroundCarry] is not null, this value is stored there.
+  /// Otherwise, the end-around carry is internally added to [sum]. This
+  /// happens when subtracting a smaller number from a larger oneusing
+  /// ones complement arithmetic.
+  Logic? get endAroundCarry => tryOutput('endAroundCarry');
 
   @protected
   Logic _sign = Logic();
 
+  /// Subtraction is happening
+  @protected
+  late final Logic? subtractIn;
+
   /// [OnesComplementAdder] constructor with an adder functor [adderGen].
   /// - Either an optional Logic [subtractIn] or a boolean [subtract] can enable
   /// subtraction, but providing both non-null will result in an exception.
-  /// - If Logic [carryOut] is provided as not null, then the end-around carry
-  ///  is not performed and is provided as value on [carryOut].
+  /// - If Logic [endAroundCarry] is provided as not null, then the end-around
+  /// carry is not performed and is provided as value on [endAroundCarry]. If
+  /// [endAroundCarry] is null, extra hardware takes care of adding the
+  /// end-around carry to [sum].
   /// - [carryIn] allows for another adder to chain into this one.
+  /// - [chainable] tells this adder to not store the [endAroundCarry] in the
+  /// sign bit as well, but to zero that to allow adders to be chained such as
+  /// for use in the [CarrySelectCompoundAdder].
   OnesComplementAdder(super.a, super.b,
       {Adder Function(Logic, Logic, {Logic? carryIn}) adderGen =
           ParallelPrefixAdder.new,
       Logic? subtractIn,
-      Logic? carryOut,
-      Logic? carryIn,
-      bool? subtract,
+      Logic? endAroundCarry,
+      super.carryIn,
+      bool subtract = false,
+      bool chainable = false,
+      String? definitionName,
       super.name = 'ones_complement_adder'})
-      : super(definitionName: 'OnesComplementAdder_W${a.width}') {
-    if (subtractIn != null) {
-      subtractIn = addInput('subtractIn', subtractIn);
-    }
-    _sign = addOutput('sign');
-    if (carryOut != null) {
-      addOutput('carryOut');
-      carryOut <= this.carryOut!;
+      : super(
+            definitionName:
+                definitionName ?? 'OnesComplementAdder_W${a.width}') {
+    if (endAroundCarry != null) {
+      addOutput('endAroundCarry');
+      endAroundCarry <= this.endAroundCarry!;
     }
-    if ((subtractIn != null) & (subtract != null)) {
+    if ((subtractIn != null) & subtract) {
```

Comment:
why not make `subtract` a `bool?`?

Reply:
If subtract is a bool?  then I have to check for both null and false for the same code.  It seems redundant.  For a hard parameter (not a Logic),  we would only use true or false and default to false; in that case we would probably not be tempted to convert our parameters to nullable throughout, so why do it here?  Adjacent to the subtractIn, it seems odd,
but against regular hard parameters, it seems reasonable.

Reply:
It seems like there are 3 options:
- this adder does not do subtraction always (`subtract = false`)
- this adder does do subtraction always (`subtract = true`)
- this adder can do addition or subtraction, based on whether `subtractIn` is driven to a 1 or 0

In the third case, it seems like either a value of `true` or `false` would be confusing.

There are other APIs i could also see as reasonable:
- a single `dynamic subtract` which must either be a `bool` or `Logic`, with default to `false`
- just a `subtractIn` and the default could be `Const(0)` (if null is provided?).  This would generate the additional SV, but synthesis would squash it down to equivalent

Reply:
This seems like an issue we should add because there are multiple places I've used this pattern (fp components, integer multiply, etc.) and it threads through them.

Another alternative would be a small class that is BoolOrSignal  to carry that configuration, rather than dynamic (I see that and I think (void*)).  

Another pattern that may be an issue that I have been using is optional Logic outputs, where we might have used a bool instead and generated the Logic output, but I chose to pass in the Logic, which generates the output and is connected to it.  Another single-bit case.

So we could decide on a common pattern for HCL and I could implement that pattern throughout.

Reply:
sure, we can split it off in another issue.  so then we need to clarify at least the documentation.

```
  /// - Either an optional Logic [subtractIn] or a boolean [subtract] can enable
  /// subtraction, but providing both non-null will result in an exception.
```

we should be explicit that `subtract == false` and `subtractIn != null` is legal and that `subtractIn` wins, if you want to keep it that way for now